ferretdb/ferretdb.go

```diff
@@ -214,21 +217,33 @@ func (f *FerretDB) MongoDBURI() string {
 	return u.String()
 }
 
-// logger is a global logger used by FerretDB.
-//
-// TODO https://github.com/FerretDB/FerretDB/issues/4014
-var logger *zap.Logger
+var (
+	loggerOnce sync.Once
+	logger     *zap.Logger
+)
 
-// Initialize the global logger there to avoid creating too many issues for zap users that initialize it in their
-// `main()` functions. It is still not a full solution; eventually, we should remove the usage of the global logger.
-//
-// TODO https://github.com/FerretDB/FerretDB/issues/4014
-func init() {
-	l := zap.ErrorLevel
-	if version.Get().DebugBuild {
-		l = zap.DebugLevel
+// getLoggerOrDefault returns the provided logger if it's not nil,
+// otherwise, it returns the global logger.
+func getLoggerOrDefault(l *zap.Logger) *zap.Logger {
+	if l != nil {
+		return l
 	}
 
-	logging.Setup(l, "console", "")
-	logger = zap.L()
+	return getGlobalLogger()
+}
+
+// getGlobalLogger retrieves or creates a global logger using
+// a loggerOnce to ensure it is created only once.
+func getGlobalLogger() *zap.Logger {
```

Comment:
for tests need 'to create a separate file' or 'update the ferretdb_test.go file of the package' or 'export functions'

I don't like any of these approaches, they remain untested 😢

Reply:
It's okay, I don't think we need them here